### ISSUE TITLE
关于模型输出

### ISSUE THREAD

**Issue by ysh000112** [ppaanngggg/layoutreader#19]
您好！请问模型输出的是具体的reading order顺序还是边界框列表中的索引呀

**Comment by ppaanngggg** [comment]
不太记得了，得看看代码了。好像是这个框应该在序列中排第几个。

**Comment by puppyapple** [comment]
> 不太记得了，得看看代码了。好像是这个框应该在序列中排第几个。

我用自己训练的模型预测时候发现一个问题，就是不论怎样random.shuffle输入的bbox的顺序（输入的bbox的ensor明显变了），得到的logits的整体分布都没有太大变化，这个是不是有些不正常？ @ppaanngggg

**Comment by ppaanngggg** [comment]
logits分布没啥变化是啥意思，没理解。那最后的结果对吗？

**Comment by puppyapple** [comment]
> logits分布没啥变化是啥意思，没理解。那最后的结果对吗？

@ppaanngggg 就是每个bbox对应的那一行的tensor的值的大小分布差不多，parse_logits得到的结果完全一模一样

**Comment by ppaanngggg** [comment]
那应该是没训好吧，loss正常吗

**Comment by puppyapple** [comment]
> 那应该是没训好吧，loss正常吗

@ppaanngggg 正常，我说的完全一样指的是对应这个输入的bbox列表而言，例如有5个bbox ABCDE，输入[B,C,D,E,A]或者[A,B,C,D,E]得到的order列表是一样的；换一个样本order结果肯定是不同的，但是同样本下打乱输入顺序的order结果总一样；
关于结果是否正确的问题，特别奇怪：
我找了一个我测试集里的样本，我再三确认了肯定不在训练数据中，然后跑了一次结果和我的label的order列表是完全一致的，shuffle了一下输入的bbox顺序，得到的依然一样，这个就不太可能是训练过拟合了。。。所以感觉很奇怪

**Comment by ppaanngggg** [comment]
> 但是同样本下打乱输入顺序的order结果总一样；

也就是说无论如何打乱，结果始终是对的？那没问题呀，训练的很好啊

**Comment by puppyapple** [comment]
> > 但是同样本下打乱输入顺序的order结果总一样；
> 
> 也就是说无论如何打乱，结果始终是对的？那没问题呀，训练的很好啊

@ppaanngggg 
举个具体的例子就是如果我输入[[100, 200, 300, 400], [200, 500, 600, 800], [10, 120, 230, 450]]，parse_logits得到了[0,2,1];
[ [200, 500, 600, 800], [100, 200, 300, 400], [10, 120, 230, 450]]，parse_logits得到的依然是[0,2,1]，我说的一样指的就是这个列表每次完全一模一样

**Comment by ppaanngggg** [comment]
那肯定不对，你要不先检查下你的训练数据格式对不对，或者训练的时候有没有打乱顺序，可能他只记住了1D position embedding

**Comment by puppyapple** [comment]
> 那肯定不对，你要不先检查下你的训练数据格式对不对，或者训练的时候有没有打乱顺序，可能他只记住了1D position embedding

@ppaanngggg 格式大概是这样的，进行了打乱：
```json
{
    "image_path": "./images/de12dec4-img_000.jpg",
    "width": 1654,
    "height": 2339,
    "source_boxes": [
        [
            1055,
            657,
            1159,
            688
        ],
        [
            707,
            2137,
            1001,
            2164
        ],
        [
            91,
            1596,
            1003,
            1793
        ],
        [
            1064,
            931,
            1554,
            1213
        ],
        [
            1055,
            853,
            1158,
            884
        ],
        [
            91,
            1018,
            1003,
            1252
        ],
        [
            91,
            853,
            1003,
            1011
        ],
        [
            91,
            1815,
            1003,
            2131
        ],
        [
            1052,
            535,
            1502,
            646
        ],
        [
            1055,
            1264,
            1208,
            1295
        ],
        [
            95,
            2137,
            443,
            2164
        ],
        [
            1053,
            694,
            1561,
            844
        ],
        [
            1500,
            198,
            1604,
            258
        ],
        [
            1055,
            484,
            1263,
            523
        ],
        [
            1054,
            1305,
            1562,
            1541
        ],
        [
            1056,
            431,
            1323,
            473
        ],
        [
            93,
            424,
            253,
            475
        ],
        [
            91,
            486,
            1003,
            644
        ],
        [
            91,
            650,
            1004,
            847
        ],
        [
            1054,
            1227,
            1403,
            1254
        ],
        [
            91,
            1260,
            1004,
            1494
        ],
        [
            91,
            1501,
            1003,
            1588
        ],
        [
            89,
            195,
            1006,
            347
        ]
    ],
    "element_types": [
        "TableCaption",
        "TableFootnote",
        "Text",
        "ImageBody",
        "ImageCaption",
        "Text",
        "Text",
        "TableBody",
        "Text",
        "Title",
        "TableFootnote",
        "TableBody",
        "Text",
        "Title",
        "List",
        "Title",
        "Title",
        "Text",
        "Text",
        "ImageFootnote",
        "Text",
        "Text",
        "Title"
    ],
    "target_index": [
        16,
        12,
        9,
        19,
        18,
        6,
        5,
        10,
        15,
        21,
        11,
        17,
        1,
        14,
        22,
        13,
        2,
        3,
        4,
        20,
        7,
        8,
        0
    ]
}
```
我不确定是不是我的shuffle逻辑有问题：
```python
random.seed(42)
if order_shuffle_rate > 0.0 and random.random() < order_shuffle_rate:
    triple_list = list(zip(source_boxes, element_types, target_index))
    random.shuffle(triple_list)
    # unzip
    source_boxes, element_types, target_index = zip(*triple_list)
```
我是直接将把元组列表进行了shuffle，然后重新zip起来的。

**Comment by puppyapple** [comment]
我的normalize 0-1000是放在datacollocator里进行的

**Comment by puppyapple** [comment]
> 不太记得了，得看看代码了。好像是这个框应该在序列中排第几个。

@ppaanngggg @jiangsufirstlove 我把shuffle逻辑改成了和你代码中的一样了，然后再训一个版本试试看；
然后输出的含义我确认了下，输出的整数列表表示按照这个列表作为索引来顺序阅读输入的bbox，例如输入是l = [A, B, C, D]，输出是[2,3,1,0]，意思就是阅读顺序应该是 l[2] l[3] l[1] l[0]，也就是CDBA

**Comment by ppaanngggg** [comment]
不是吧，我记得应该是表示A排在第2个，B排在第三个这样子，所以应该是D，C，A，B

**Comment by puppyapple** [comment]
> 不是吧，我记得应该是表示A排在第2个，B排在第三个这样子，所以应该是D，C，A，B

@ppaanngggg 我用你的那个逻辑生成了数据之后专门画了图的，按我说的那种画出来的是正确的图；我还专门去确认了下MinerU里吗的调用逻辑

**Comment by ppaanngggg** [comment]
好吧，可能我记错了，你确定代码正确就行

**Comment by puppyapple** [comment]
> 好吧，可能我记错了，你确定代码正确就行

@ppaanngggg  改了下shuffle方式问题还是一样的，我猜测是不是因为block级别的数据每个页面元素太少了，因此模型很容易过拟合到一些特定位置预测特定index的pattern上，这也解释了输出总是一样的原因。。。

**Comment by ppaanngggg** [comment]
应该不是。。我们内部有个模型是可以用的

**Comment by puppyapple** [comment]
> 应该不是。。我们内部有个模型是可以用的

@ppaanngggg emmm那就奇怪，我仔细研究了下数据，没发现有什么不正常的地方，😭

**Comment by puppyapple** [comment]
> > 应该不是。。我们内部有个模型是可以用的
> 
> [@ppaanngggg](https://github.com/ppaanngggg) emmm那就奇怪，我仔细研究了下数据，没发现有什么不正常的地方，😭

@ppaanngggg 请问你们的block数据构建和训练上有做什么差异化的处理吗？

**Comment by ppaanngggg** [comment]
完全一样，只是换了私有的数据

**Comment by puppyapple** [comment]
> 完全一样，只是换了私有的数据

@ppaanngggg  有一个小细节就是bbox的坐标需要在1000以内，我看代码里似乎没有显式进行归一化？

**Comment by ppaanngggg** [comment]
需要处理到1000以内的，我通常按照图像宽高都scale到1000来算
